### PR TITLE
Feature/venom wireless tests

### PR DESCRIPTION
**Pull request by JeGoi**
# Description
Test for wireless

# Impacts
Test for wireless 

# Delete branch after merge
YES Please

# Checklist
- [x] Finish yaml tests files
- [x] Finish tests documentation
- [x] Run tests with success

### CODE REVIEW

**Comment by nqb** [comment]
Pipeline: https://gitlab.com/inverse-inc/packetfence/-/pipelines/268350582

File: addons/vagrant/inventory/hosts [hunk 1]
```diff
@@ -143,18 +143,19 @@ all:
...
         forward_mode: 'route'
+        netmask: 255.255.255.0
 
       - name: 'inline'
         subnet: '172.17.6.0/24'
         forward_mode: 'route'
-
-
-      
+        netmask: 255.255.255.0
```
Comment (nqb) [inline addons/vagrant/inventory/hosts:161]: `netmask `is not necessary if `/XX` is specified after `subnet`

File: t/venom/nodes/common/check_wlan1_int_status_down.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+name: Check Wlan1 interface status on wireless01
```
Comment (nqb) [inline t/venom/nodes/common/check_wlan1_int_status_down.yml]: I will add "down" in title

File: t/venom/nodes/common/check_wlan1_int_status_down.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+version: 2
+testcases:
+- name: check_wlan1_int_status_on_wireless01
```
Comment (nqb) [inline t/venom/nodes/common/check_wlan1_int_status_down.yml]: same here

File: t/venom/nodes/common/check_wlan1_int_status_up.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+name: Check Wlan1 interface status on wireless01
```
Comment (nqb) [inline t/venom/nodes/common/check_wlan1_int_status_up.yml]: I will add "up" in title

File: t/venom/nodes/common/check_wlan1_int_status_up.yml [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+version: 2
+testcases:
+- name: check_wlan1_int_status_on_wireless01
```
Comment (nqb) [inline t/venom/nodes/common/check_wlan1_int_status_up.yml]: I will add "up" in title

File: t/venom/nodes/wireless_mac_auth/run_wpasupplicant.yml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+name: Run wpasupplicant for Wireless Secure
```
Comment (nqb) [inline t/venom/nodes/wireless_mac_auth/run_wpasupplicant.yml]:
```suggestion
name: Run wpasupplicant for Wireless Open
```

File: t/venom/nodes/wireless_mac_auth/run_wpasupplicant.yml [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+version: 2
+testcases:
+- name: run_wpasupplicant_wireless_secure
```
Comment (nqb) [inline t/venom/nodes/wireless_mac_auth/run_wpasupplicant.yml]:
```suggestion
- name: run_wpasupplicant_wireless_open
```

File: t/venom/vars/all.yml [hunk 8]
```diff
@@ -94,17 +102,40 @@ cumulus_switch.id1: '{{.switch01_mgmt_ip}}'
...
+wireless_switch_group.description: Wireless switch group
+wireless_switch_group.radiusSecret: anHardSharedSecret
+
+# Wireless Switch
+wireless_switch.id1: '{{.wireless01_mgmt_ip}}'
+wireless_switch.id2: '{{.wireless01.dot1x_interface.mac}}'
+wireless_switch.description: wireless01
+
+################################################################################
+# dot1x_eap_peap test suite specific variables
```
Comment (nqb) [inline t/venom/vars/all.yml]:
```suggestion
# eap_peap test suites specific variables
```

File: t/venom/vars/all.yml [hunk 9]
```diff
@@ -122,27 +153,76 @@ dot1x_eap_peap.profiles.wired.filters.connection_type: Ethernet-EAP
...
+
 ################################################################################
 ## backup db and restore test suite specific variables
 #################################################################################
 # Create user for db
 backup_db_and_restore.user: adminvenom
 backup_db_and_restore.backup_dir: /root/backup
 
+################################################################################
+## Roles mac auth test suite specific variables
```
Comment (nqb) [inline t/venom/vars/all.yml]:
```suggestion
## mac auth test suites specific variables
```

File: t/venom/vars/all.yml [hunk 10]
```diff
@@ -57,7 +59,13 @@ switch01.dot1x_interface.mac: 44:38:39:00:00:12
 node01_mgmt_ip: 172.17.17.251
 node01_eth1_mac_address: 00:03:00:11:11:01
 node01_eth1_mac_address_url_encoded: 00%3A03%3A00%3A11%3A11%3A01
-nodes_test_suite_dir: /src/t/venom/nodes
+
+# wireless01 variables
+wireless01_mgmt_ip: 172.17.17.210
+wireless01_wlan1_mac_address: 02:00:00:00:01:00
+wireless01_wlan1_mac_address_url_encoded: 02%3A00%3A00%3A00%3A01%3A00
+wireless01.dot1x_interface.id: eth2
```
Comment (nqb) [inline t/venom/vars/all.yml]: This variable is not used anywhere.

File: addons/vagrant/inventory/group_vars/wireless/hostapd.yml [hunk 11]
```diff
@@ -3,42 +3,46 @@
...
-    hostapd__ctrl_interface_group: '0'
+    hostapd__ctrl_interface_group: 0
     hostapd__driver: 'nl80211'
     hostapd__hw_mode: 'g'
     hostapd__interface: 'wlan0'
-    hostapd__logger_syslog: '1'
-    hostapd__macaddr_acl: '2'
+    hostapd__logger_syslog: 1
+    hostapd__macaddr_acl: 2
     hostapd__nas_identifier: 'HostAP-PF'
```
Comment (nqb) [inline addons/vagrant/inventory/group_vars/wireless/hostapd.yml]:
```suggestion
    hostapd__nas_identifier: '{{ ansible_fqdn }}'
```
That's way, we know request is coming from `wireless01`.

File: addons/vagrant/inventory/group_vars/wireless/hostapd.yml [hunk 12]
```diff
@@ -3,42 +3,46 @@
...
     hostapd__driver: 'nl80211'
-    hostapd__ieee8021x: '1'
+    hostapd__ieee8021x: 1
     hostapd__hw_mode: 'g'
     hostapd__interface: 'wlan0'
-    hostapd__logger_syslog: '1'
-    hostapd__macaddr_acl: '2'
+    hostapd__logger_syslog: 1
+    hostapd__macaddr_acl: 2
     hostapd__nas_identifier: 'HostAP-PF'
```
Comment (nqb) [inline addons/vagrant/inventory/group_vars/wireless/hostapd.yml]:
```suggestion
    hostapd__nas_identifier: '{{ ansible_fqdn }}'
```
That's way, we know request is coming from `wireless01`.

File: t/venom/pfservers/wireless_mac_auth/teardown/20_check_node01_deleted.yml [hunk 13]
```diff
@@ -0,0 +1 @@
+../../common/check_node01_deleted.yml
```
Comment (nqb) [inline t/venom/pfservers/wireless_mac_auth/teardown/20_check_node01_deleted.yml]: Can be removed

File: t/venom/pfservers/wireless_mac_auth/TESTSUITE.md [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+1. Configure MAC authentication and dynamic VLAN on dot1x interface on
+   wireless01: launch hostapd_open
+1. Wait some time to let hostapd_open running
+   wireless01: launch wpa_supplicant open
+1. Wait some time to let RADIUS request be sent by switch01 and handled by
+   PacketFence server (`mab_activation_delay` + 20 seconds). wireless01 also needs to handle answer.
+1. Check RADIUS audit log for wireless01 wlan1
+1. Check VLAN assigned to wireless01 *on* wlan0 this the tag id
+
+## Teardown steps
```
Comment (nqb) [inline t/venom/pfservers/wireless_mac_auth/TESTSUITE.md:28]:
```suggestion
## Teardown steps
1. Unregister wireless01 (wlan1)
```

File: t/venom/pfservers/wireless_mac_auth/TESTSUITE.md [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+- delete_windows=1m
+1. Restart `pfcron` to take change into account
+1. Create a node with MAC address of wireless01 (wlan1) : 02:00:00:00:01:00
+- assign the role headless_device without unreg date
+- add a note
+1. Create connection profile with specific filter Wireless-802.11-NoEAP
+1. Configure MAC authentication and dynamic VLAN on dot1x interface on
+   wireless01: launch hostapd_open
+1. Wait some time to let hostapd_open running
+   wireless01: launch wpa_supplicant open
```
Comment (nqb) [inline t/venom/pfservers/wireless_mac_auth/TESTSUITE.md]:
```suggestion
1. Enable WPA supplicant on wireless01
```

File: t/venom/pfservers/wireless_mac_auth/TESTSUITE.md [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+1. Wait some time to let RADIUS request be sent by switch01 and handled by
+   PacketFence server (`mab_activation_delay` + 20 seconds). wireless01 also needs to handle answer.
```
Comment (nqb) [inline t/venom/pfservers/wireless_mac_auth/TESTSUITE.md]:
```suggestion
1. Wait some time to let RADIUS request be sent by wireless01 and handled by
   PacketFence server. wireless01 also needs to handle answer.
```

File: t/venom/pfservers/global_config/15_create_network_devices.yml [hunk 17]
```diff
@@ -141,3 +141,166 @@ testcases:
...
+        "wsTransport": null,
+        "wsUser": null
+      }
+    headers:
+      "Authorization": "{{.pfserver_token}}"
+      "Content-Type": "application/json"
+    assertions:
+      - result.statuscode ShouldEqual 201
+
+# workaround to make accounting working:
```
Comment (nqb) [inline t/venom/pfservers/global_config/15_create_network_devices.yml]: This workaround is not necessary for wireless01 because NAS-IP-Address equals wireless01 management IP address in Accounting messages.

Currently, wireless01 is only created by MAC address like an AP with controllerIP defined. 

We can keep things like this but remove that comment.

File: t/venom/pfservers/wireless_mac_auth/00_enable_wireless01_cleanup_task.yml [hunk 18]
```diff
@@ -0,0 +1,37 @@
+name: Enable wireless_cleanup maintenance task and change the delete_window
```
Comment (nqb) [inline t/venom/pfservers/wireless_mac_auth/00_enable_wireless01_cleanup_task.yml]:
```suggestion
name: Enable node_cleanup maintenance task and change the delete_window
```

Comment (nqb) [reply t/venom/pfservers/wireless_mac_auth/00_enable_wireless01_cleanup_task.yml]: Filename needs to be adapted too.

File: t/venom/pfservers/wireless_mac_auth/02_restart_pfcron_service.yml [hunk 19]
```diff
@@ -0,0 +1,14 @@
+name: Restart pfcron service
```
Comment (nqb) [inline t/venom/pfservers/wireless_mac_auth/02_restart_pfcron_service.yml]: We can replace this by a symlink to `../common/restart_pfcron_service.yml`

File: t/venom/pfservers/wireless_mac_auth/05_create_wireless01.yml [hunk 20]
```diff
@@ -0,0 +1,62 @@
+name: Create wireless01 with the mac 02:00:00:00:01:00 and role headless_device
```
Comment (nqb) [inline t/venom/pfservers/wireless_mac_auth/05_create_wireless01.yml]:
```suggestion
name: Create wireless01 with the mac 02:00:00:00:01:00 (wlan1 interface) and role headless_device
```

File: t/venom/pfservers/wireless_mac_auth/20_enable_wpa_supplicant.yml [hunk 21]
```diff
@@ -0,0 +1,9 @@
+name: Enable wpa_supplicant on wireless01
+version: 2
+testcases:
+- name: enable_dynamic_vlan_on_wireless01
```
Comment (nqb) [inline t/venom/pfservers/wireless_mac_auth/20_enable_wpa_supplicant.yml]:
```suggestion
- name: enable_wpa_supplicant_on_wireless01
```

File: t/venom/pfservers/wireless_mac_auth/teardown/07_disable_mac_auth_int.yml [hunk 22]
```diff
@@ -0,0 +1,9 @@
+name: Enable MAC Authentication on dot1x interface on wireless01
```
Comment (nqb) [inline t/venom/pfservers/wireless_mac_auth/teardown/07_disable_mac_auth_int.yml]:
```suggestion
name: Disable MAC Authentication on dot1x interface on wireless01
```

**Review by nqb — CHANGES_REQUESTED** [review]
### RADIUS Disconnect using `hostapd` 

hostapd configuration need to be updated for secure and open SSID to allow PacketFence to disconnect nodes when we unreg devices on PacketFence.

Directives:

```
radius_das_port=3799
radius_das_client=PF_MGMT_IP anHardSharedSecret
radius_das_require_event_timestamp=0
radius_das_require_message_authenticator=0
```

Tested on wireless_mac_auth scenario, it works.

According to Ansible wireless collection, it looks like you just need to update inventory because hostapd template is already ready.

### wireless_mac_auth

#### Teardown: removal of wireless01 node

Teardown failed to remove `wireless01` node because you don't wait enough time to run node_cleanup task.

Take a look on how it's done on wired_mac_auth scenario using `12_sleep_delete_windows.yml`

File: addons/vagrant/wireless/Vagrantfile [hunk 23]
```diff
@@ -30,8 +30,29 @@ Vagrant.configure("2") do |config|
...
+                     :libvirt__dhcp_enabled => false,
+                     :libvirt__forward_mode => networks[1]['forward_mode']
+      # link iso_network (vlan 3)
+      srv.vm.network "private_network",
+                     :mac => "44:34:22:11:00:03",
+                     :libvirt__network_name => networks[2]['name'],
+                     :ip => "172.17.3.100",
+                     :netmask => networks[2]['netmask'],
+                     :libvirt__dhcp_enabled => false,
+                     :libvirt__forward_mode => networks[2]['forward_mode']
```
Comment (nqb) [inline addons/vagrant/wireless/Vagrantfile]: Is it still necessary ?

File: t/venom/pfservers/wireless_dot1x_eap_peap/teardown/05_disable_wireless_dot1x_eap_peap_int.yml [hunk 24]
```diff
@@ -0,0 +1,9 @@
+name: Enable MAC Authentication on dot1x interface on wireless01
```
Comment (nqb) [inline t/venom/pfservers/wireless_dot1x_eap_peap/teardown/05_disable_wireless_dot1x_eap_peap_int.yml]:
```suggestion
name: Disable 802.1X on dot1x interface on wireless01
```

File: t/venom/pfservers/wireless_dot1x_eap_peap/teardown/05_disable_wireless_dot1x_eap_peap_int.yml [hunk 25]
```diff
@@ -0,0 +1,9 @@
...
+version: 2
+testcases:
+- name: enable_mac_auth_on_wireless01
```
Comment (nqb) [inline t/venom/pfservers/wireless_dot1x_eap_peap/teardown/05_disable_wireless_dot1x_eap_peap_int.yml]:
```suggestion
- name: disable_dot1x_on_wireless01
```

File: t/venom/switches/common/check_wlan0.100_int_status.yml [hunk 26]
```diff
@@ -0,0 +1,10 @@
+name: Check wlan0.2 interface status on wireless01
```
Comment (nqb) [inline t/venom/switches/common/check_wlan0.100_int_status.yml]: Is this file still necessary ?

File: t/venom/pfservers/wireless_dot1x_eap_peap/35_enable_dot1x_eap_peap_int.yml [hunk 27]
```diff
@@ -0,0 +1,9 @@
+name: Enable MAC Authentication on dot1x interface on wireless01
```
Comment (nqb) [inline t/venom/pfservers/wireless_dot1x_eap_peap/35_enable_dot1x_eap_peap_int.yml]:
```suggestion
name: Enable 802.1X on dot1x interface on wireless01
```

File: t/venom/pfservers/wireless_dot1x_eap_peap/35_enable_dot1x_eap_peap_int.yml [hunk 28]
```diff
@@ -0,0 +1,9 @@
...
+version: 2
+testcases:
+- name: enable_mac_auth_on_wireless01
```
Comment (nqb) [inline t/venom/pfservers/wireless_dot1x_eap_peap/35_enable_dot1x_eap_peap_int.yml]:
```suggestion
- name: enable_dot1x_on_wireless01
```

**Review by nqb — CHANGES_REQUESTED** [review]
### `check_wlan1` are not used

Following files doesn't seem used:

```
./nodes/common/check_wlan1_int_status_up.yml
./nodes/common/check_wlan1_int_status_down.yml
```

**Comment by nqb** [comment]
Last things:
- [X] I will update documentation to reflect changes introduced by this PR (topology, IP addresses and MAC)
- [X] My test were done using Vagrant box `debian/buster64` 10.4.0 version. CI will use 10.20210228.1 with a change on the way network is managed (systemd-networkd is disabled)

=> JeGoi did its tests using 10.20210228.1 box.

- [X] I will add notes in TESTSUITE.md about:
  * stopping of hostapd will remove wlan0.XXX interfaces
  * during wireless_dot1x_eap_peap scenario, we receive a MAC Auth request **before** 802.1X request

File: addons/vagrant/wireless/Vagrantfile (hunk 23 above)
Comment (JeGoi) [reply addons/vagrant/wireless/Vagrantfile]: Since we are not able to map the hostapd dynamic vlan on register vlan, no it is not necessary.
I will remove it.
Thanks

File: addons/vagrant/inventory/hosts (hunk 1 above)
Comment (JeGoi) [reply addons/vagrant/inventory/hosts:161]: Yes but it is used when you set a specific ip. It is what I used for reg and iso network.
I guess we can let them here even if we remove iso and reg interface in wireless01.

File: t/venom/switches/common/check_wlan0.100_int_status.yml (hunk 26 above)
Comment (JeGoi) [reply t/venom/switches/common/check_wlan0.100_int_status.yml]: No, you are right. I use a generic based on variable now.

File: addons/vagrant/inventory/hosts (hunk 1 above)
Comment (nqb) [reply addons/vagrant/inventory/hosts:161]: OK, I understand. It should be possible to set up an IP using X.X.X.X/24 with `ip`

Comment (JeGoi) [reply addons/vagrant/inventory/hosts:161]: Hum, Not sure it will work. For example: X.X.X.X/24 with `ip` is not used like that in https://github.com/inverse-inc/packetfence/blob/devel/addons/vagrant/pfservers/Vagrantfile#L33

**Comment by JeGoi** [comment]
> ### `check_wlan1` are not used
> Following files doesn't seem used:
> 
> ```
> ./nodes/common/check_wlan1_int_status_up.yml
> ./nodes/common/check_wlan1_int_status_down.yml
> ```

Hum, I don't know if they can be used during venom tests. If the timing is not accurate it could be a source of error and at the same time, it could be useful to check if it is there or not.

**Comment by nqb** [comment]
> ### wireless_mac_auth
> #### Teardown: removal of wireless01 node
> Teardown failed to remove `wireless01` node because you don't wait enough time to run node_cleanup task.

I'm not sure it will work with your fix. I suggest we use: `../../common/sleep_delete_windows.yml` in place of:
- 12_15s_sleep_time.yml
- 14_30s_sleep_time.yml
